packages/core/src/conversation/ConversationService/ConversationService.ts

```diff
@@ -388,50 +430,170 @@ export class ConversationService extends TypedEventEmitter<Events> {
+      if (keyPackages && keyPackages.length > 0) {
+        await this.mlsService.addUsersToExistingConversation(groupId, keyPackages);
 
-      //We store the info when user was added (and key material was created), so we will know when to renew it
-      await this.mlsService.resetKeyMaterialRenewal(groupId);
-    }
+        //We store the info when user was added (and key material was created), so we will know when to renew it
+        await this.mlsService.resetKeyMaterialRenewal(groupId);
+      }
 
-    return {
-      conversation,
-      failedToAdd: keysClaimingFailures,
-    };
+      return {
+        conversation,
+        failedToAdd: keysClaimingFailures,
+      };
+    } catch (error) {
+      if (MLSService.isBrokenMLSConversationError(error)) {
+        if (!shouldRetry) {
+          this.logger.warn("Tried to add users to MLS conversation but it's still broken after reset", error);
+          throw error;
+        }
+        this.logger.warn("Tried to add users to MLS conversation but it's broken, resetting the conversation", error);
+        return this.handleBrokenMLSConversation(conversationId, newGroupId =>
+          this.addUsersToMLSConversation({qualifiedUsers, groupId: newGroupId, conversationId, shouldRetry: false}),
+        );
+      }
+      throw error;
+    }
   }
 
   public async removeUsersFromMLSConversation({
     groupId,
     conversationId,
     qualifiedUserIds,
-  }: RemoveUsersParams): Promise<Conversation> {
-    const clientsToRemove = await this.apiClient.api.user.postListClients({qualified_users: qualifiedUserIds});
+    shouldRetry = true,
+  }: RemoveUsersParams & {shouldRetry?: boolean}): Promise<Conversation> {
+    try {
+      const clientsToRemove = await this.apiClient.api.user.postListClients({qualified_users: qualifiedUserIds});
 
-    const fullyQualifiedClientIds = mapQualifiedUserClientIdsToFullyQualifiedClientIds(
-      clientsToRemove.qualified_user_map,
-    );
+      const fullyQualifiedClientIds = mapQualifiedUserClientIdsToFullyQualifiedClientIds(
+        clientsToRemove.qualified_user_map,
+      );
 
-    await this.mlsService.removeClientsFromConversation(groupId, fullyQualifiedClientIds);
+      await this.mlsService.removeClientsFromConversation(groupId, fullyQualifiedClientIds);
 
-    //key material gets updated after removing a user from the group, so we can reset last key update time value in the store
-    await this.mlsService.resetKeyMaterialRenewal(groupId);
+      // key material gets updated after removing a user from the group, so we can reset last key update time value in the store
+      await this.mlsService.resetKeyMaterialRenewal(groupId);
 
-    return await this.getConversation(conversationId);
+      return await this.getConversation(conversationId);
+    } catch (error) {
+      if (MLSService.isBrokenMLSConversationError(error)) {
+        if (!shouldRetry) {
```

Comment:
we just log, what will happen then?

Reply:
then we let the call to this method throw an error that reaches surface of the webapp, this is the current behavior also.